lib/rules/no-shadow.js

```diff
@@ -57,6 +67,70 @@ module.exports = {
             allow: (context.options[0] && context.options[0].allow) || []
         };
 
+        /**
+         * @callback CheckNodeCallback
+         * @param {ASTNode} node A node to check.
+         */
+
+        /**
+         * Traverses AST from a given node to find a node that verifies checkNodeCB condition.
+         * @param {ASTNode} node a node to get.
+         * @param {CheckNodeCallback} checkNodeCB a callback that checks whether or not the node verifies its condition or not.
+         * @returns {ASTNode} the node that verifies checkNodeCB condition.
+         */
+        function getNode(node, checkNodeCB) {
```

Comment:
I think the names aren’t quite right here. I think this is a search operation, so maybe the names should reflect that, like:

```suggestion
        function findSelfOrAncestor(node, match) {
```